Resolve sources in stack traces that are generated on remote JavaScript runtime

We generate sourcemaps for each test bundle which means that in the browser where an exception occurs, the sources are resolved. However, when the same exception is transmitted across the network, logging it out to the console does not resolve the sources because 

1. it isn't an `Error` class
2. the stack trace was generated by code on a remote JavaScript vm and so the sourcemap is not intrinsically associated with the code.


We need to first figure out how to do this period, possibly by manually loading the sourcemap, and and then figure out how to integrate it into our output.

### Example

```
 AssertionError: expected 2 == 3
|       at _callee$ (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:66693:23)
|       at tryCatch (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:196:23)
|       at Generator.invoke [as _invoke] (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:413:28)
|       at Generator.next (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:249:27)
|       at asyncGeneratorStep (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:861:28)
|       at _next (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:883:13)
|       at http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:890:11
|       at new Promise (<anonymous>)
|       at http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:879:16
|       at _assertEqual (http://localhost:24005/manifest-44d22a51a12a34996032ee8496870689be93fe591ea90818f31fe310be8dfdfc.js:6670
```

--- a/.changeset/source-maps.md
+++ b/.changeset/source-maps.md
@@ -1,0 +1,8 @@
+---
+"@bigtest/agent": "minor"
+"@bigtest/server": "minor"
+"@bigtest/cli": "minor"
+"@bigtest/suite": "minor"
+---
+
+Resolve source maps in error stack traces for better debugging

--- a/.changeset/variables.md
+++ b/.changeset/variables.md
@@ -1,0 +1,5 @@
+---
+"@bigtest/client": "minor"
+---
+
+Add support for variables in GraphQL queries

--- a/packages/agent/app/lane.ts
+++ b/packages/agent/app/lane.ts
@@ -1,16 +1,9 @@
-import { timebox } from './timebox';
-
-import { Agent } from '../shared/agent';
-import { TestImplementation, ErrorDetails, Context as TestContext } from '@bigtest/suite';
 import { Operation, fork } from 'effection';
+import { TestImplementation, Context as TestContext } from '@bigtest/suite';
 
-const serializeError: (error: ErrorDetails) => ErrorDetails = ({ message, fileName, lineNumber, columnNumber, stack }) => ({
-  message,
-  fileName,
-  lineNumber,
-  columnNumber,
-  stack
-});
+import { timebox } from './timebox';
+import { serializeError } from './serialize-error';
+import { Agent } from '../shared/agent';
 
 export function *runLane(testRunId: string, agent: Agent, test: TestImplementation, path: string[]): Operation<void> {
   return yield runLaneSegment(testRunId, agent, test, {}, path.slice(1), []);
@@ -50,7 +43,7 @@ function *runLaneSegment(testRunId: string, agent: Agent, test: TestImplementati
           type: 'step:result',
           status: 'failed',
           timeout: false,
-          error: serializeError(error),
+          error: yield serializeError(error),
           path: stepPath
         });
       }
@@ -71,7 +64,7 @@ function *runLaneSegment(testRunId: string, agent: Agent, test: TestImplementati
           agent.send({ testRunId, type: 'assertion:result', status: 'ok', path: assertionPath });
         } catch(error) {
           console.error('[agent] assertion failed', assertion, error);
-          agent.send({ testRunId, type: 'assertion:result', status: 'failed', error: serializeError(error), path: assertionPath });
+          agent.send({ testRunId, type: 'assertion:result', status: 'failed', error: yield serializeError(error), path: assertionPath });
         }
       });
     }

--- a/packages/agent/app/serialize-error.ts
+++ b/packages/agent/app/serialize-error.ts
@@ -8,6 +8,7 @@ interface SourceLocation {
   name?: string;
   line?: number;
   column?: number;
+  sourceLine?: string;
   sourceFile: {
     path: string;
   };
@@ -24,6 +25,7 @@ export function *resolveStackFrames(stackFrames: StackFrame[]): Operation<ErrorS
         return {
           name: location.name || stackFrame.functionName,
           fileName: stackFrame.fileName,
+          code: location.sourceLine,
           line: stackFrame.lineNumber,
           column: stackFrame.columnNumber,
           source: {

--- a/packages/agent/app/serialize-error.ts
+++ b/packages/agent/app/serialize-error.ts
@@ -1,0 +1,40 @@
+import * as getSource from 'get-source';
+import * as ErrorStackParser from 'error-stack-parser';
+import { Operation } from 'effection';
+import { ErrorDetails } from '@bigtest/suite';
+
+interface SourceLocation {
+  name?: string;
+  line?: number;
+  column?: number;
+  sourceFile: {
+    path: string;
+  };
+}
+
+export function *serializeError(error: Error): Operation<ErrorDetails> {
+  let parsedError = yield Promise.all(ErrorStackParser.parse(error).map(async (stackFrame) => {
+    if(stackFrame.fileName && stackFrame.lineNumber && stackFrame.columnNumber) {
+      let source = await getSource.async(stackFrame.fileName);
+      let location: SourceLocation = await source.resolve({ line: stackFrame.lineNumber, column: stackFrame.columnNumber });
+      return {
+        name: location.name || stackFrame.functionName,
+        fileName: stackFrame.fileName,
+        line: stackFrame.lineNumber,
+        column: stackFrame.columnNumber,
+        source: {
+          fileName: location.sourceFile.path,
+          line: location.line,
+          column: location.column
+        }
+      };
+    } else {
+      return { fileName: stackFrame.fileName, line: stackFrame.lineNumber, column: stackFrame.columnNumber };
+    }
+  }));
+  return {
+    name: error.name,
+    message: error.message,
+    stack: parsedError
+  };
+}

--- a/packages/agent/app/serialize-error.ts
+++ b/packages/agent/app/serialize-error.ts
@@ -1,7 +1,8 @@
 import * as getSource from 'get-source';
 import * as ErrorStackParser from 'error-stack-parser';
-import { Operation } from 'effection';
-import { ErrorDetails } from '@bigtest/suite';
+import { StackFrame } from 'error-stack-parser';
+import { Operation, spawn, join } from 'effection';
+import { ErrorDetails, ErrorStackFrame } from '@bigtest/suite';
 
 interface SourceLocation {
   name?: string;
@@ -12,29 +13,59 @@ interface SourceLocation {
   };
 }
 
-export function *serializeError(error: Error): Operation<ErrorDetails> {
-  let parsedError = yield Promise.all(ErrorStackParser.parse(error).map(async (stackFrame) => {
-    if(stackFrame.fileName && stackFrame.lineNumber && stackFrame.columnNumber) {
-      let source = await getSource.async(stackFrame.fileName);
-      let location: SourceLocation = await source.resolve({ line: stackFrame.lineNumber, column: stackFrame.columnNumber });
-      return {
-        name: location.name || stackFrame.functionName,
-        fileName: stackFrame.fileName,
-        line: stackFrame.lineNumber,
-        column: stackFrame.columnNumber,
-        source: {
-          fileName: location.sourceFile.path,
-          line: location.line,
-          column: location.column
-        }
-      };
+export function *resolveStackFrames(stackFrames: StackFrame[]): Operation<ErrorStackFrame[]> {
+  let children = [];
+
+  for(let stackFrame of stackFrames) {
+    children.push(yield spawn(function*() {
+      if(stackFrame.fileName && stackFrame.lineNumber && stackFrame.columnNumber) {
+        let source = yield getSource.async(stackFrame.fileName);
+        let location: SourceLocation = yield source.resolve({ line: stackFrame.lineNumber, column: stackFrame.columnNumber });
+        return {
+          name: location.name || stackFrame.functionName,
+          fileName: stackFrame.fileName,
+          line: stackFrame.lineNumber,
+          column: stackFrame.columnNumber,
+          source: {
+            fileName: location.sourceFile.path,
+            line: location.line,
+            column: location.column
+          }
+        };
+      } else {
+        return { fileName: stackFrame.fileName, line: stackFrame.lineNumber, column: stackFrame.columnNumber };
+      }
+    }));
+  }
+
+  let results = [];
+
+  for(let child of children) {
+    results.push(yield join(child));
+  }
+
+  return results;
+}
+
+export function *serializeError(error: unknown): Operation<ErrorDetails> {
+  try {
+    if(error instanceof Error) {
+      let stackFrames = ErrorStackParser.parse(error);
+
+      let resolvedStackFrames: ErrorStackFrame[];
+      try {
+        resolvedStackFrames = yield resolveStackFrames(stackFrames);
+      } catch {
+        resolvedStackFrames = [];
+      }
+
+      return { name: error.name, message: error.message, stack: resolvedStackFrames };
     } else {
-      return { fileName: stackFrame.fileName, line: stackFrame.lineNumber, column: stackFrame.columnNumber };
+      return {
+        message: `${error}`
+      }
     }
-  }));
-  return {
-    name: error.name,
-    message: error.message,
-    stack: parsedError
-  };
+  } catch {
+    return { message: "unknown error" }
+  }
 }

--- a/packages/agent/package.json
+++ b/packages/agent/package.json
@@ -46,7 +46,9 @@
     "@bigtest/globals": "^0.6.0",
     "@effection/events": "^0.7.7",
     "bowser": "^2.9.0",
-    "effection": "^0.7.0"
+    "effection": "^0.7.0",
+    "error-stack-parser": "^2.0.6",
+    "get-source": "^2.0.11"
   },
   "volta": {
     "node": "12.16.0",

--- a/packages/cli/src/formatters/lines.ts
+++ b/packages/cli/src/formatters/lines.ts
@@ -26,6 +26,9 @@ function formatEvent(event: RunResultEvent) {
         if(stackFrame.name) {
           result += `@ ${stackFrame.name}`;
         }
+        if(stackFrame.code) {
+          result += `\n|        > ${stackFrame.code.trim()}`
+        }
       }
     }
   }

--- a/packages/cli/src/formatters/lines.ts
+++ b/packages/cli/src/formatters/lines.ts
@@ -1,9 +1,5 @@
 import { StreamingFormatter, Counts, RunResultEvent, icon } from '../format-helpers';
 
-function filterStack(text: string): string {
-  return text.split('\n').filter((l) => !l.match(/__bigtest/)).join('\n');
-}
-
 function formatFooterCounts(label: string, counts: Counts): string {
   return [
     `${label}:`.padEnd(14),
@@ -19,16 +15,21 @@ function formatEvent(event: RunResultEvent) {
     result += ' ' + event.path.slice(1).join(' -> ');
   }
   if(event.error) {
-    result += '\n' + prefixLines(event.error.stack ? filterStack(event.error.stack) : `Error: ${event.error.message}`, '|   ')
+    result += ["\n|    ERROR:", event.error.name, event.error.message].filter(e => e).join(' ');
+    if(event.error.stack) {
+      for(let stackFrame of event.error.stack) {
+        let location = stackFrame.source || stackFrame;
+        result += `\n|      `
+        if(location.fileName) {
+          result += `${location.fileName}:${location.line || 0}:${location.column || 0} `;
+        }
+        if(stackFrame.name) {
+          result += `@ ${stackFrame.name}`;
+        }
+      }
+    }
   }
   return result;
-}
-
-function prefixLines(text: string, prefix: string) {
-  return text
-    .split('\n')
-    .map((l) => prefix + l)
-    .join('\n')
 }
 
 const formatter: StreamingFormatter = {

--- a/packages/cli/src/query.ts
+++ b/packages/cli/src/query.ts
@@ -2,7 +2,7 @@ import { ResultStatus, ErrorDetails } from '@bigtest/suite';
 
 export function run() {
   return `
-    subscription($showInternal: Boolean! = true, $showDependencies: Boolean! = true) {
+    subscription($showInternalStackTrace: Boolean! = true, $showDependenciesStackTrace: Boolean! = true) {
       event: run {
         type
         status
@@ -11,7 +11,7 @@ export function run() {
         path
         error {
           message
-          stack(showInternal: $showInternal, showDependencies: $showDependencies) {
+          stack(showInternal: $showInternalStackTrace, showDependencies: $showDependenciesStackTrace) {
             name
             fileName
             line

--- a/packages/cli/src/query.ts
+++ b/packages/cli/src/query.ts
@@ -2,7 +2,7 @@ import { ResultStatus, ErrorDetails } from '@bigtest/suite';
 
 export function run() {
   return `
-    subscription($showInternalStackTrace: Boolean! = true, $showDependenciesStackTrace: Boolean! = true) {
+    subscription($showInternalStackTrace: Boolean! = true, $showDependenciesStackTrace: Boolean! = true, $showStackTraceCode: Boolean! = true) {
       event: run {
         type
         status
@@ -14,6 +14,7 @@ export function run() {
           stack(showInternal: $showInternalStackTrace, showDependencies: $showDependenciesStackTrace) {
             name
             fileName
+            code @include(if: $showStackTraceCode)
             line
             column
             source {

--- a/packages/cli/src/query.ts
+++ b/packages/cli/src/query.ts
@@ -1,8 +1,8 @@
-import { ResultStatus } from '@bigtest/suite';
+import { ResultStatus, ErrorDetails } from '@bigtest/suite';
 
 export function run() {
   return `
-    subscription {
+    subscription($showInternal: Boolean! = true, $showDependencies: Boolean! = true) {
       event: run {
         type
         status
@@ -11,10 +11,17 @@ export function run() {
         path
         error {
           message
-          fileName
-          lineNumber
-          columnNumber
-          stack
+          stack(showInternal: $showInternal, showDependencies: $showDependencies) {
+            name
+            fileName
+            line
+            column
+            source {
+              fileName
+              line
+              column
+            }
+          }
         }
         timeout
       }
@@ -27,13 +34,7 @@ export type RunResultEvent = {
   testRunId: string;
   status?: ResultStatus;
   path?: string[];
-  error?: {
-    message: string;
-    fileName?: string;
-    lineNumber?: string;
-    columnNumber?: string;
-    stack?: string;
-  };
+  error?: ErrorDetails;
   timeout?: boolean;
 }
 

--- a/packages/cli/src/run-test.ts
+++ b/packages/cli/src/run-test.ts
@@ -25,7 +25,7 @@ export function* runTest(config: ProjectOptions, formatter: StreamingFormatter):
     }
   };
 
-  let subscription = yield client.subscription(query.run(), { showDependencies: false, showInternal: false });
+  let subscription = yield client.subscription(query.run(), { showDependenciesStackTrace: false, showInternalStackTrace: false });
   let stepCounts = { ok: 0, failed: 0, disregarded: 0 };
   let assertionCounts = { ok: 0, failed: 0, disregarded: 0 };
   let testRunStatus: ResultStatus | undefined;

--- a/packages/cli/src/run-test.ts
+++ b/packages/cli/src/run-test.ts
@@ -25,7 +25,11 @@ export function* runTest(config: ProjectOptions, formatter: StreamingFormatter):
     }
   };
 
-  let subscription = yield client.subscription(query.run(), { showDependenciesStackTrace: false, showInternalStackTrace: false });
+  let subscription = yield client.subscription(query.run(), {
+    showDependenciesStackTrace: false,
+    showInternalStackTrace: false,
+    showStackTraceCode: false
+  });
   let stepCounts = { ok: 0, failed: 0, disregarded: 0 };
   let assertionCounts = { ok: 0, failed: 0, disregarded: 0 };
   let testRunStatus: ResultStatus | undefined;

--- a/packages/cli/src/run-test.ts
+++ b/packages/cli/src/run-test.ts
@@ -10,23 +10,22 @@ import { StreamingFormatter } from './format-helpers';
 export function* runTest(config: ProjectOptions, formatter: StreamingFormatter): Operation<void> {
 
   let uri = `ws://localhost:${config.port}`;
-  
+
   let client: Client = yield function*() {
     try {
       return yield Client.create(uri);
     } catch (e) {
       if (e.name === 'NoServerError') {
-        throw new MainError({ 
+        throw new MainError({
           exitCode: 1,
           message: `Could not connect to BigTest server on ${uri}. Run "bigtest server" to start the server.`
         });
       }
       throw e;
-    }  
+    }
   };
 
-  let subscription = yield client.subscription(query.run());
-
+  let subscription = yield client.subscription(query.run(), { showDependencies: false, showInternal: false });
   let stepCounts = { ok: 0, failed: 0, disregarded: 0 };
   let assertionCounts = { ok: 0, failed: 0, disregarded: 0 };
   let testRunStatus: ResultStatus | undefined;

--- a/packages/cli/test/cli.test.ts
+++ b/packages/cli/test/cli.test.ts
@@ -140,6 +140,7 @@ describe('@bigtest/cli', function() {
         expect(child.stdout?.output).toContain("✓ [step]       Failing Test -> first step")
         expect(child.stdout?.output).toContain("✓ [assertion]  Failing Test -> check the thing")
         expect(child.stdout?.output).toContain("⨯ [step]       Failing Test -> child -> child second step")
+        expect(child.stdout?.output).toContain("test/fixtures/failing.test.ts:14")
         expect(child.stdout?.output).toContain("⨯ FAILURE")
       });
     });

--- a/packages/client/src/client.ts
+++ b/packages/client/src/client.ts
@@ -4,7 +4,7 @@ import { resource, Operation, spawn } from 'effection';
 import { Mailbox } from '@bigtest/effection';
 import { on, once } from '@effection/events';
 
-import { Message, isErrorResponse, isDataResponse, isDoneResponse } from './protocol';
+import { Variables, Message, isErrorResponse, isDataResponse, isDoneResponse } from './protocol';
 import { NoServerError } from './errors';
 
 let responseIds = 0;
@@ -19,7 +19,7 @@ export class Client {
 
     yield spawn(function* detectStartupError(): Operation<void> {
       let [error] = yield once(socket, 'error');
-      
+
       if (isErrorEvent(error)) {
         throw new NoServerError(`Could not connect to server at ${url}`);
       } else {
@@ -44,25 +44,25 @@ export class Client {
     return res;
   }
 
-  *query(source: string): Operation {
-    let subscription = yield this.send("query", source, false);
+  *query(source: string, variables?: Variables): Operation {
+    let subscription = yield this.send("query", source, false, variables);
     return yield subscription.receive();
   }
 
-  *mutation(source: string): Operation {
-    let subscription = yield this.send("mutation", source);
+  *mutation(source: string, variables?: Variables): Operation {
+    let subscription = yield this.send("mutation", source, false, variables);
     return yield subscription.receive();
   }
 
-  *liveQuery(source: string): Operation<Mailbox> {
-    return yield this.send("query", source, true);
+  *liveQuery(source: string, variables?: Variables): Operation<Mailbox> {
+    return yield this.send("query", source, true, variables);
   }
 
-  *subscription(source: string): Operation<Mailbox> {
-    return yield this.send("subscription", source);
+  *subscription(source: string, variables?: Variables): Operation<Mailbox> {
+    return yield this.send("subscription", source, false, variables);
   }
 
-  private *send(type: string, source: string, live = false): Operation<Mailbox> {
+  private *send(type: string, source: string, live = false, variables?: Variables): Operation<Mailbox> {
     let mailbox = new Mailbox();
     let { socket } = this;
 
@@ -71,7 +71,7 @@ export class Client {
 
       let responseId = `${responseIds++}`; //we'd want a UUID to avoid hijacking?
 
-      socket.send(JSON.stringify({ [type]: source, live, responseId}));
+      socket.send(JSON.stringify({ [type]: source, live, responseId, variables }));
 
       while (true) {
         let { value: [event] } = yield messages.next();

--- a/packages/client/src/protocol.ts
+++ b/packages/client/src/protocol.ts
@@ -1,7 +1,10 @@
+export type Variables = Record<string, unknown>;
+
 export interface Message {
   query?: string;
   mutation?: string;
   subscription?: string;
+  variables?: Variables;
   responseId?: string;
 }
 

--- a/packages/server/src/command-server.ts
+++ b/packages/server/src/command-server.ts
@@ -10,7 +10,7 @@ import { GraphqlContext, SpawnContext } from './schema/context';
 import { Atom } from '@bigtest/atom';
 import { OrchestratorState } from './orchestrator/state';
 
-import { Message, Response, QueryMessage, MutationMessage, SubscriptionMessage, isQuery, isMutation, isSubscription } from '@bigtest/client';
+import { Variables, Message, Response, QueryMessage, MutationMessage, SubscriptionMessage, isQuery, isMutation, isSubscription } from '@bigtest/client';
 
 export type CommandMessage = { status: "ready" } | { type: "run"; id: string };
 
@@ -47,10 +47,10 @@ export function* createCommandServer(options: CommandServerOptions): Operation {
  * Run the query or mutation in `source` against the orchestrator
  * state contained in `state`
  */
-function* graphql(source: string, options: CommandServerOptions, state: OrchestratorState): Operation {
+function* graphql(source: string, variables: Variables | undefined, options: CommandServerOptions, state: OrchestratorState): Operation {
   let context: SpawnContext = yield spawn(undefined);
   let opts = graphqlOptions(context, options, state);
-  return yield executeGraphql({...opts, contextValue: opts.context, source });
+  return yield executeGraphql({...opts, contextValue: opts.context, source, variableValues: variables });
 }
 
 /**
@@ -76,7 +76,7 @@ function handleMessage(options: CommandServerOptions): (socket: Socket) => Opera
   }
 
   function* handleMutation(message: MutationMessage, socket: Socket): Operation {
-    let result: Response = yield graphql(message.mutation, options, options.atom.get());
+    let result: Response = yield graphql(message.mutation, message.variables, options, options.atom.get());
     result.responseId = message.responseId;
     yield socket.send(result);
   }
@@ -87,7 +87,8 @@ function handleMessage(options: CommandServerOptions): (socket: Socket) => Opera
     let result = yield executeGraphqlSubscription({
       schema,
       document: parseGraphql(message.subscription),
-      contextValue: new GraphqlContext(context, options.atom, options.delegate)
+      contextValue: new GraphqlContext(context, options.atom, options.delegate),
+      variableValues: message.variables
     });
 
     if(isAsyncIterator(result)) {
@@ -113,7 +114,7 @@ function handleMessage(options: CommandServerOptions): (socket: Socket) => Opera
   }
 
   function* publishQueryResult(message: QueryMessage, state: OrchestratorState, socket: Socket): Operation {
-    let result: Response = yield graphql(message.query, options, state);
+    let result: Response = yield graphql(message.query, message.variables, options, state);
     result.responseId = message.responseId;
     yield socket.send(result);
   }

--- a/packages/server/src/manifest-server.ts
+++ b/packages/server/src/manifest-server.ts
@@ -7,12 +7,17 @@ interface ManifestServerOptions {
   delegate: Mailbox;
   dir: string;
   port: number;
+  proxyPort: number;
 };
 
 export function* createManifestServer(options: ManifestServerOptions): Operation {
   let app = express();
 
-  app.raw.use(staticMiddleware(options.dir));
+  app.raw.use(staticMiddleware(options.dir, {
+    setHeaders(res) {
+      res.setHeader('Access-Control-Allow-Origin', `http://localhost:${options.proxyPort}`);
+    }
+  }));
 
   yield app.listen(options.port);
 

--- a/packages/server/src/orchestrator.ts
+++ b/packages/server/src/orchestrator.ts
@@ -85,6 +85,7 @@ export function* createOrchestrator(options: OrchestratorOptions): Operation {
     delegate: manifestServerDelegate,
     dir: manifestDistDir,
     port: options.project.manifest.port,
+    proxyPort: options.project.proxy.port,
   }));
 
   yield fork(createManifestGenerator({

--- a/packages/server/src/schema.ts
+++ b/packages/server/src/schema.ts
@@ -5,9 +5,11 @@ import {
   mutationType,
   subscriptionField,
   stringArg,
+  booleanArg,
   makeSchema,
   enumType,
 } from "@nexus/schema";
+import { ErrorStackFrame } from '@bigtest/suite';
 
 export const schema = makeSchema({
   typegenAutoConfig: {
@@ -288,12 +290,53 @@ export const schema = makeSchema({
     }),
     objectType({
       name: "Error",
+      rootTyping: {
+        name: "ErrorDetails",
+        path: "@bigtest/suite"
+      },
       definition(t) {
         t.string("message");
+        t.list.field("stack", {
+          type: "ErrorStackFrame",
+          nullable: true,
+          args: {
+            showInternal: booleanArg({ required: false, default: true }),
+            showDependencies: booleanArg({ required: false, default: true }),
+          },
+          async resolve({ stack }, { showInternal, showDependencies }) {
+            if(stack) {
+              return stack.filter((frame: ErrorStackFrame) => {
+                if(!showInternal && frame.fileName?.match(/__bigtest/)) {
+                  return false;
+                }
+                if(!showDependencies && frame.source?.fileName?.match(/node_modules/)) {
+                  return false;
+                }
+                return true;
+              });
+            } else {
+              return null;
+            }
+          }
+        });
+      }
+    }),
+    objectType({
+      name: "ErrorStackFrame",
+      definition(t) {
+        t.string("name", { nullable: true });
         t.string("fileName", { nullable: true });
-        t.int("lineNumber", { nullable: true });
-        t.int("columnNumber", { nullable: true });
-        t.string("stack", { nullable: true });
+        t.int("line", { nullable: true });
+        t.int("column", { nullable: true });
+        t.field("source", { type: "ErrorStackFrameSource", nullable: true });
+      }
+    }),
+    objectType({
+      name: "ErrorStackFrameSource",
+      definition(t) {
+        t.string("fileName", { nullable: true });
+        t.int("line", { nullable: true });
+        t.int("column", { nullable: true });
       }
     })
   ]

--- a/packages/server/src/schema.ts
+++ b/packages/server/src/schema.ts
@@ -326,6 +326,7 @@ export const schema = makeSchema({
       definition(t) {
         t.string("name", { nullable: true });
         t.string("fileName", { nullable: true });
+        t.string("code", { nullable: true });
         t.int("line", { nullable: true });
         t.int("column", { nullable: true });
         t.field("source", { type: "ErrorStackFrameSource", nullable: true });

--- a/packages/suite/src/interfaces.ts
+++ b/packages/suite/src/interfaces.ts
@@ -123,5 +123,6 @@ export interface ErrorStackLocation {
 
 export interface ErrorStackFrame extends ErrorStackLocation {
   name?: string;
+  code?: string;
   source?: ErrorStackLocation;
 }

--- a/packages/suite/src/interfaces.ts
+++ b/packages/suite/src/interfaces.ts
@@ -110,9 +110,18 @@ export interface AssertionResult extends Node {
 }
 
 export interface ErrorDetails {
+  name?: string;
   message: string;
+  stack?: ErrorStackFrame[];
+}
+
+export interface ErrorStackLocation {
   fileName?: string;
-  lineNumber?: number;
-  columnNumber?: number;
-  stack?: string;
+  line?: number;
+  column?: number;
+}
+
+export interface ErrorStackFrame extends ErrorStackLocation {
+  name?: string;
+  source?: ErrorStackLocation;
 }

--- a/packages/webdriver/package.json
+++ b/packages/webdriver/package.json
@@ -31,7 +31,7 @@
     "@effection/fetch": "^0.1.2",
     "@effection/node": "^0.8.0",
     "abort-controller": "^3.0.0",
-    "chromedriver": "~83.0.0",
+    "chromedriver": "~85.0.0",
     "effection": "^0.7.0",
     "geckodriver": "^1.19.1",
     "node-fetch": "^2.6.0"

